File: src/UE4_Assets/FMeshUVChannelInfo.java
/**
 * 
 */
package UE4_Assets;

import UE4.FArchive;
import annotation.FixedArraySize;
import annotation.BooleanZ;
import annotation.Float32;
import annotation.Serializable;
import annotation.UeExclude;
import lombok.Data;

/**
 * @author FunGames
 *
 */
@Data
@Serializable
public class FMeshUVChannelInfo {

	@UeExclude public static final int TEXSTREAM_MAX_NUM_UVCHANNELS = 4;

	@BooleanZ private boolean bInitialized;
	@BooleanZ private boolean bOverrideDensities;
	@FixedArraySize(TEXSTREAM_MAX_NUM_UVCHANNELS) @Float32 private float[] localUVDensities;
	
	public FMeshUVChannelInfo(FArchive Ar) throws ReadException {
		bInitialized = Ar.readBoolean();
		bOverrideDensities = Ar.readBoolean();
		localUVDensities = new float[TEXSTREAM_MAX_NUM_UVCHANNELS];
		for(int i=0;i<TEXSTREAM_MAX_NUM_UVCHANNELS; i++) {
			localUVDensities[i] = Ar.readFloat32();
		}
	}

}

File: src/main/kotlin/me/fungames/jfortniteparse/ue4/assets/objects/meshes/FMeshUVChannelInfo.kt
package me.fungames.jfortniteparse.ue4.assets.objects.meshes

import me.fungames.jfortniteparse.ue4.assets.UProperty
import me.fungames.jfortniteparse.ue4.assets.UStruct
import me.fungames.jfortniteparse.ue4.reader.FArchive
import me.fungames.jfortniteparse.ue4.writer.FArchiveWriter

internal const val TEXSTREAM_MAX_NUM_UVCHANNELS = 4

@UStruct
class FMeshUVChannelInfo {
    @JvmField @UProperty("bInitialized")
    var initialized: Boolean
    @JvmField @UProperty("bOverrideDensities")
    var overrideDensities: Boolean
    @JvmField @UProperty("LocalUVDensities", arrayDim = TEXSTREAM_MAX_NUM_UVCHANNELS)
    var localUVDensities: Array<Float>

    constructor(Ar: FArchive) {
        initialized = Ar.readBoolean()
        overrideDensities = Ar.readBoolean()
        localUVDensities = Array(TEXSTREAM_MAX_NUM_UVCHANNELS) { Ar.readFloat32() }
    }

    fun serialize(Ar: FArchiveWriter) {
        Ar.writeBoolean(initialized)
        Ar.writeBoolean(overrideDensities)
        localUVDensities.forEach { Ar.writeFloat32(it) }
    }

    constructor(initialized: Boolean, overrideDensities: Boolean, localUVDensities: Array<Float>) {
        this.initialized = initialized
        this.overrideDensities = overrideDensities
        this.localUVDensities = localUVDensities
    }

    constructor() : this(false, false, emptyArray())
}